### PR TITLE
noVNC console integration

### PR DESCRIPTION
**Pull request by davidjumani**
## Description
This feature provides novnc console for virtual machines.

It is developed based on noVNC 1.1.0
It includes changes in cloudstack :
(1) add a new websocket proxy to interact with the noVNC console
(2) a ConsoleProxyClient to read / write data from the vm vnc to the websocket.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?
(1) Works on kvm / ubuntu / centos
(2) Works on xs 75
(3) Works on VMware 6.5

### CODE REVIEW

**Comment by davidjumani** [comment]
@ustcweizhou I've created this to use a java websocket proxy which supports client metrics reporting. Could you provide your comments

**Comment by weizhouapache** [comment]
@davidjumani cooooool.
I am currently busy on some other stuff. I will look at it when I am free.
@rhtyd has some ideas. maybe he can help you.

**Comment by davidjumani** [comment]
@rhtyd It's all good from my end, any comments ?

**Comment by nvazquez** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/servlet/ConsoleProxyServlet.java [hunk 1]
```diff
@@ -478,7 +481,12 @@ private String composeConsoleAccessUrl(String rootUrl, VirtualMachine vm, HostVO
             param.setClientTunnelSession(parsedHostInfo.third());
         }
 
-        sb.append("/ajax?token=" + encryptor.encryptObject(ConsoleProxyClientParam.class, param));
+        if (param.getHypervHost() != null || !ConsoleProxyManager.NoVncConsoleDefault.value()) {
+            sb.append("/ajax?token=" + encryptor.encryptObject(ConsoleProxyClientParam.class, param));
+        } else {
+            sb.append("/resource/noVNC/vnc_lite.html?port=" + ConsoleProxyManager.DEFAULT_NOVNC_PORT + "&token="
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/servlet/ConsoleProxyServlet.java]: Why not using `vnc.html` which provides more functionalities? Briefly tested invoking it and got this extra functionalities:

image

File: server/src/main/java/com/cloud/consoleproxy/ConsoleProxyManager.java [hunk 2]
```diff
@@ -31,9 +33,14 @@
     public static final int DEFAULT_PROXY_URL_PORT = 80;
     public static final int DEFAULT_PROXY_SESSION_TIMEOUT = 300000;        // 5 minutes
 
+    public static final int DEFAULT_NOVNC_PORT = 8080;
+
     public static final String ALERT_SUBJECT = "proxy-alert";
     public static final String CERTIFICATE_NAME = "CPVMCertificate";
 
+    public static final ConfigKey<Boolean> NoVncConsoleDefault = new ConfigKey<Boolean>("Advanced", Boolean.class, "novnc.console.default", "true",
+        "If true, noVNC console will be default console for virtual machines", true);
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/consoleproxy/ConsoleProxyManager.java]: I think it should be disabled by default at least until all hypervisors tested working properly

Comment (davidjumani) [reply server/src/main/java/com/cloud/consoleproxy/ConsoleProxyManager.java]: I've tested it on KVM, VMware and Xen and it works fine!

File: server/src/main/java/com/cloud/servlet/ConsoleProxyServlet.java (hunk 1 above)
Comment (davidjumani) [reply server/src/main/java/com/cloud/servlet/ConsoleProxyServlet.java]: Wanted a simpler, minimalist UI, can add this when we get multi lingual keyboard working

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by vladimirpetrov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by vladimirpetrov** [comment]
@blueorangutan  help

**Comment by rhtyd** [comment]
@blueorangutan help

**Comment by vladimirpetrov** [comment]
@blueorangutan help

**Comment by vladimirpetrov** [comment]
@blueorangutan package

**Comment by vladimirpetrov** [comment]
@blueorangutan test

**Comment by davidjumani** [comment]
Verified with multi keyboard layouts! Have to configure the appropriate keyboard layout in the guest VM
keyboards

**Comment by vladimirpetrov** [comment]
@blueorangutan package

**Comment by vladimirpetrov** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
@blueorangutan test

**Comment by andrijapanicsb** [comment]
Travis log length exceeded - otherwise LGMT

Merging based on 2 x LGTMs, manual testing by 2 persons (Vladimir and me) and the regression test results passing 100%.

**Comment by DaanHoogland** [comment]
@davidjumani please have a look at https://builds.apache.org/job/cloudstack-pr-analysis/10377/artifact/target/rat.txt or equivalent of any new PR. it seems we missed some licensing checks on this PR.

**Pull request merged**